### ISSUE TITLE
Linear gradients may be rendered differently between femtovg and Qt

### ISSUE THREAD

**Issue by tronical** [sixtyfpsui/sixtyfps#313]
The following minimal test-case produces different rendering with femtovg and Qt:

```60
T := Window {
  preferred-width: 640px;
  preferred-height: 480px;
  Rectangle {
    background: @linear-gradient(0, red, yellow, green, transparent);
  }
}
```

Ideally they are rendered the same.

Screenshots to illustrate:

| femtovg | Qt |
|-|-|
| <img width="708" alt="femtovg" src="https://user-images.githubusercontent.com/1486/125287053-c514db80-e31c-11eb-83fe-67941eb0f38e.png"> | <img width="752" alt="qt" src="https://user-images.githubusercontent.com/1486/125287077-cc3be980-e31c-11eb-986e-15d73fe879f1.png"> |

**Comment by nyanpasu64** [comment]
My guess as to the issue is that you're not using premultiplied alpha for blending, so halfway between `rgba(0, 0, 0, 0)` and `rgba(0, 255, 0, 1)` is blended as "draw halfway between black and green, with alpha halfway between 0 and 100%" rather than "draw green with 50% alpha".

~~Another possibility for improving gradients is using gamma-corrected blending for gradients.~~ This produces more consistent brightnesses (halfway through a gradient is the same brightness as a checkerboard, blending from red to green isn't darker in the middle). The downside is that it makes black-white linear gradients lighter in the middle than expected. Unfortunately this shouldn't be implemented, because browsers blend in sRGB gamma space rather than linear.

**Comment by tronical** [comment]
Thank you @nyanpasu64!! What you’re saying about the lack of premultiplied alpha makes sense just looking at the fragment shader in femtovg. I’ll take a look on Monday.

**Comment by ogoffart** [comment]
We have more renderer now.

The Qt and software renderer agree,  and Skia and Femtovg also agree

<img width="971" alt="image" src="https://github.com/slint-ui/slint/assets/959326/dc701b06-ae9b-4fe5-bd0c-edc3a0e3042c">

But who is right?

**Comment by ogoffart** [comment]
I think Qt and software is right.   Skia and FemtoVG show some "gray" color which seems coming from the pixel value of the transparent pixel. (which is black with 0 alpha)

**Comment by tronical** [comment]
Yes, Qt and software renderer are right IMO - they're interpolating pre-multiplied colors and there's no grey.

**Comment by tronical** [comment]
Related femtovg issue: https://github.com/femtovg/femtovg/issues/16

**Comment by matthargett** [comment]
not sure if it will match Qt now, but femtovg [#326](https://github.com/femtovg/femtovg/pull/326) implemented/fixed the remaining linear gradient conformance gaps identified by the SVG and Canvas WPT tests. so, maybe the intent of this issue is satisfied? let me know!
